sound/soc/sof/trace.c

```diff
@@ -582,7 +595,7 @@ void snd_sof_free_trace(struct snd_sof_dev *sdev)
 	if (!sdev->dtrace_is_supported)
 		return;
 
-	snd_sof_release_trace(sdev);
+	snd_sof_release_trace(sdev, 0);
```

Comment:
false? I think we should pass what the final state we want instead of true/false. For example if you want to stop and release, you pass SOF_DTRACE_DISABLED and SOF_DTRACE_STOPPED otherwise.

Reply:
> false? I think we should pass what the final state we want instead of true/false. For example if you want to stop and release, you pass SOF_DTRACE_DISABLED and SOF_DTRACE_STOPPED otherwise.

@ujfalusi What do you think?

Reply:
@ranj063 @ujfalusi Let me explain why I prefer ```void snd_sof_release_trace(struct snd_sof_dev *sdev, bool only_stop)```. Actually, I'm also satisfied with @ranj063 's first suggestion ```snd_sof_release_trace(struct snd_sof_dev *sdev, bool release) ```. And after talking with @ujfalusi , we thought 'only_stop' may be better. Anyway, the 2 name are both OK to me. And I'm prefer ```bool only_stop``` or ``` bool release``` to  ```sof_dtrace_state```. This is because:
1. snd_sof_release_trace() is just executing what the caller tells it to do. This function should not determine what to do itself. But maybe the caller need different behavior in the same pm state in different scenarios in the future, for example releasing the dma trace even it is in D0 state. So this function should be flexible and easy. Based on this theory, snd_sof_release_trace() should not judge what to do itself, but let the caller tells it what to do. 
2. In the ```snd_sof_free_trace```, the target state is not D3. We didn't set the audio state to D3 anywhere in the free. So we should set the parameter to D0 if we use  ```sof_dtrace_state``` for the precise description of the power state. But this will make the function do the wrong thing. (In the free case, we must set the state to D3). But if we set the parameter to D3, this is misleading.

Reply:
@libinyang, from the pm.c on suspend you should pass the pm_state to the dtrace, likely by adding a new function: `snd_sof_trace_suspend(sdev, pm_state)` (align the name with the rest of external trace API).
Internally in trace.c you could then convert this pm_state to whatever bool you end up using for `snd_sof_release_trace()`, but this function must be `static` to trace.c.
For the resume part I would add a `snd_sof_trace_resume(sdev)` for symmetry and sanity. Sure, if we were in D3 we will do a full startup (sans host memory and buffer allocation, debugfs and whatnot), but at the end it we are resuming the dtrace functionality after a suspend.

Reply:
passing the pm_state makes a lot more sense than a bool. That way the trace logic makes the decision on what the final state should be during release

Reply:
This can be indeed considered from both directions @libinyang , but if you consider the caller here, e.g. pm.c, it's not really ideal place either to decide what exactly should happen when trace is released. And if you pass a boolean, then it's really pm.c as the caller here that needs to know details about how trace works. We want to decouple SOF core (like pm.c) from clients (like trace.c). DMA trace is still a bit of a special case, but basicly same/similar principles should apply. So in the end snd_sof_release_trace() call should be similar to sof_suspend_clients() and for that we do have:
"sof_suspend_clients(sdev, pm_state);". I.e. the idea is that the client (trace, probe, audio, injector) knows best what actions to take given core is moving to D3 or D0i.

Reply:
@kv2019i , You comments make sense to me. Yes, if we want to make trace to be a client, we should use the decoupled implementation, and let trace decide what to do. And if so, @ujfalusi 's suggestion looks good to me. BTW, we will really make the trace to be a client finally, right?

I will refine the patch and send it out as a RFC without full test. The full test takes to much time. After we all agree with the new PR, I will do the full test.

Reply:
The second parameter for `snd_sof_release_trace()` is bool, so pass `false`